editor/utils/heartbeat.js

```diff
@@ -0,0 +1,165 @@
+import store from '../store';
+import {
+	getEditedPostTitle,
+	getEditedPostExcerpt,
+	getEditedPostContent,
+	getCurrentPostId,
+	isPostAutosaveDirty,
+} from '../store/selectors';
+import {
+	toggleAutosave,
+	removeNotice,
+	toggleNetworkIsConnected,
+	showDisconnectionNotice,
+	resetAutosave,
+	updateAutosave,
+	updateAutosaveStatusMessage,
+} from '../store/actions';
+
+import { compact } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { doAction } from '@wordpress/hooks';
+
+const DISCONNECTION_NOTICE_ID = 'DISCONNECTION_NOTICE_ID';
+
+/**
+ * Set up the heartbeat functionality for Gutenberg.
+ */
+export function setupHeartbeat() {
+	const $document = jQuery( document );
+
+	/**
+	 * Configure heartbeat to refresh the wp-api nonce, keeping the editor authorization intact.
+	 *
+	 * @todo update the _wpnonce used by autosaves.
+	 */
+	$document.on( 'heartbeat-tick', ( event, response ) => {
+		if ( response[ 'rest-nonce' ] ) {
+			window.wpApiSettings.nonce = response[ 'rest-nonce' ];
+		}
+	} );
+
+	/**
+	 * Configure Heartbeat autosaves.
+	 */
+	const { dispatch, getState } = store;
+
+	let compareString;
+	let lastCompareString;
+
+	/**
+	 * Autosave 'save' function that pulls content from Gutenberg state. Based on `wp.autosave.save`.
+	 *
+	 * @since 1.9.0
+	 *
+	 * @returns {Object|bool} postData The autosaved post data to send, or false if no autosave is needed.
+	 */
+	const save = function() {
+		// Bail early if autosaving is suspended or saving is blocked.
+		if ( wp.autosave.isSuspended || wp.autosave._blockSave ) {
+			return false;
+		}
+
+		// Check if its time for another autosave.
+		if ( ( new Date() ).getTime() < wp.autosave.nextRun ) {
+			return false;
+		}
+
+		// Get the current editor state and compute the compare string (title::excerpt::content).
+		const state = getState();
+
+		const toSend = {
+			post_title: getEditedPostTitle( state ),
+			post_excerpt: getEditedPostExcerpt( state ),
+			content: getEditedPostContent( state ),
+			post_id: getCurrentPostId( state ),
+		};
+
+		// Store the current editor values into the state autosave.
+		dispatch( resetAutosave( toSend ) );
+
+		// If the autosave is clean, no need to save.
+		if ( ! isPostAutosaveDirty( state ) ) {
```

Comment:
@youknowriad this checks the isDirty status of state.autosave, if the autosave isn't dirty we don't autosave